src/Microsoft.ML.Data/Model/ModelOperationsCatalog.cs

```diff
@@ -45,13 +45,34 @@ protected SubCatalogBase(ModelOperationsCatalog owner)
         /// <param name="stream">A writeable, seekable stream to save to.</param>
         public void Save(ITransformer model, Stream stream) => model.SaveTo(Environment, stream);
 
+        public void Save<TSource>(IDataLoader<TSource> model, Stream stream)
```

Comment:
This I think this is not correct... we need to save *both* the loader (or failing that, input schema) as well as the transform model to a stream. Not two separate things.

Reply:
I recall the mistake that was made with predictor models and transform models in entry-points, where we viewed these as single things saved to a stream, as opposed to having to capture the entire pipeline. This mistake was inherited by ML.NET v0.1, which led to all sorts of grief when people realized, yes, saving the loader is *really freakin' important*. So I think we're just making the same mistake again. A model isn't just a transform model, or a loader, or an input schema, it I think needs to capture all of them, and I think we need a change a bit more fundamental than merely adding a few methods here and there.

---
In reply to: [262638162](https://github.com/dotnet/machinelearning/pull/2850#discussion_r262638162) [](ancestors = 262638162)